### PR TITLE
Add speed benchmark examples

### PR DESCRIPTION
**Pull request by wangxingjun778**
1. Add speed_benchmark_transformer.py in examples.
2. Add speed_benchmark_vllm.py in examples.
3. Add README.md and README_zh.md for speed benchmark.

### CODE REVIEW

**Comment by wangxingjun778** [comment]
Refer to #1054

File: examples/speed-benchmark/requirements/perf_vllm.txt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+vllm==0.6.3.post1
+torch==2.4.0
+flash_attn==2.6.3
```
Comment (jklj077) [inline examples/speed-benchmark/requirements/perf_vllm.txt:3]: no need for flash_attn now. vllm uses the built in vllm.flash_attn.

File: examples/speed-benchmark/speed_benchmark_transformer.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+import os
+import time
+import torch
+import pandas as pd
+import json
+from transformers.trainer_utils import set_seed
+
+"""
+Qwen2.5 Speed Benchmark for transformer(pt) inference.
```
Comment (jklj077) [inline examples/speed-benchmark/speed_benchmark_transformer.py:11]: transformers(pt)

File: examples/speed-benchmark/speed_benchmark_transformer.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+"""
+
+
+class SpeedBenchmarkTransformer:
```
Comment (jklj077) [inline examples/speed-benchmark/speed_benchmark_transformer.py:15]: SpeedBenchmarkTransformers

the library name is `transformers`

File: examples/speed-benchmark/speed_benchmark_transformer.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    BATCH_SIZE = 1
+    USE_FLASH_ATTN = True
+    COMMENT = 'default'
+    DEVICE_MAP = 'auto'
+    TORCH_DTYPE = 'auto'
+    GENERATE_LENGTH_PER_EXPERIMENT = 2048
+    OVERWRITE_RESULT = False
+    DUMMY_INPUT = '我'
+
+    def __init__(self, model_id_or_path, use_modelscope: bool = True, outputs_dir: str = 'outputs/transformer'):
```
Comment (jklj077) [inline examples/speed-benchmark/speed_benchmark_transformer.py:27]: outputs/transformers

File: examples/speed-benchmark/speed_benchmark_transformer.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
+# Copyright (c) Alibaba, Inc. and its affiliates.
```
Comment (jklj077) [inline examples/speed-benchmark/speed_benchmark_transformer.py:1]: please black or ruff this file

File: examples/speed-benchmark/speed_benchmark_vllm.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
+# Copyright (c) Alibaba, Inc. and its affiliates.
```
Comment (jklj077) [inline examples/speed-benchmark/speed_benchmark_vllm.py:1]: please format and sort import this file

File: examples/speed-benchmark/speed_benchmark_transformer.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+                                     f"{model_id_or_path_str}"
+                                     f"_context_length-{context_length}_{timestamp}.csv")
+        out_dir = os.path.dirname(out_file)
+        os.makedirs(out_dir, exist_ok=True)
+        self.save_result(data, out_file)
+
+        return out_file
+
+    @staticmethod
+    def save_result(data: dict, out_file: str) -> None:
```
Comment (jklj077) [inline examples/speed-benchmark/speed_benchmark_transformer.py:122]: i wonder if the whole purpose of the pandas requirement is just for saving csv.

File: examples/speed-benchmark/speed_benchmark_transformer.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+        df.to_csv(out_file, index=False)
+        print(f'Results saved to {out_file}')
+
+
+def main():
+
+    import argparse
+
+    # Parse args
+    parser = argparse.ArgumentParser(description='Speed benchmark for transformer(pt) deployment')
```
Comment (jklj077) [inline examples/speed-benchmark/speed_benchmark_transformer.py:133]: i think you could just use those help messages for the readme

File: examples/speed-benchmark/README.md [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+Parameters:
+
+    `--model_id_or_path`: Model ID or path, optional values refer to the Model Resources section.  
+    `--context_length`: Input length in tokens; optional values are 1, 6144, 14336, 30720, 63488, 129024; Refer to the `Qwen2.5 SpeedBenchmark`.  
+    `--gpus`: Number of GPUs to use, e.g., 0,1.  
+    `--use_modelscope`: Whether to use ModelScope; if False, HuggingFace is used; default is True.  
+    `--outputs_dir`: Output directory; default is outputs/transformer.  
+
+
+#### 3.2 Inference using vLLM
```
Comment (jklj077) [inline examples/speed-benchmark/README.md:47]: we should add the note of HF_ENDPOINT here.

File: examples/speed-benchmark/README.md [hunk 10]
```diff
@@ -0,0 +1,75 @@
+## Speed Benchmark
+
+This document introduces the speed benchmark testing process for the Qwen2.5 series models (original and quantized models). For detailed reports, please refer to the [Qwen2.5 SpeedBenchmark](../../docs/source/benchmark/speed_benchmark.rst)
```
[the remark this answers is not in the document]
Comment (wangxingjun778) [reply examples/speed-benchmark/README.md:3]: done

File: examples/speed-benchmark/requirements/perf_transformer.txt [hunk 11]
```diff
@@ -0,0 +1,9 @@
+torch==2.3.1
+transformers==4.46.0
+auto_gptq==0.7.1
+autoawq==0.2.6
+flash_attn==2.5.8
```
[the remark this answers is not in the document]
Comment (wangxingjun778) [reply examples/speed-benchmark/requirements/perf_transformer.txt:5]: done

File: examples/speed-benchmark/README.md [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+For models hosted on HuggingFace, please refer to [Qwen2.5 Collections-HuggingFace](https://huggingface.co/collections/Qwen/qwen25-66e81a666513e518adb90d9e)
+
+For models hosted on ModelScope, please refer to [Qwen2.5 Collections-ModelScope](https://modelscope.cn/collections/Qwen25-dbc4d30adb768)
+
+### 2. Environment Installation
+
+For inference using HuggingFace transformers:
+
+```shell
+pip install -r requirements/perf_transformer.txt
```
[the remark this answers is not in the document]
Comment (wangxingjun778) [reply examples/speed-benchmark/README.md:20]: done

File: examples/speed-benchmark/README.md [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+```
+
+
+### 3. Run experiments
+
+#### 3.1 Inference using HuggingFace Transformers
+
+```shell
+python speed_benchmark_transformer.py --model_id_or_path Qwen/Qwen2.5-0.5B-Instruct --context_length 1 --gpus 0 --use_modelscope --outputs_dir outputs/transformer
```
[the remark this answers is not in the document]
Comment (wangxingjun778) [reply examples/speed-benchmark/README.md]: Add `--generate_length` for transformers and vllm, set 2048 by default.

[the remark this answers is not in the document]
Comment (wangxingjun778) [reply examples/speed-benchmark/README.md]: Added desc of `--generate_length` in readme

File: examples/speed-benchmark/requirements/perf_transformer.txt [hunk 14]
```diff
@@ -0,0 +1,9 @@
...
+pandas
+modelscope[framework]
+accelerate
+optimum
```
[the remark this answers is not in the document]
Comment (wangxingjun778) [reply examples/speed-benchmark/requirements/perf_transformer.txt]: set optimum>=1.20.0

File: examples/speed-benchmark/requirements/perf_transformer.txt [hunk 15]
```diff
@@ -0,0 +1,9 @@
+torch==2.3.1
+transformers==4.46.0
+auto_gptq==0.7.1
```
[the remark this answers is not in the document]
Comment (jklj077) [reply examples/speed-benchmark/requirements/perf_transformer.txt]: If that's the case, I think we should make a separate requirements file for autoawq.

**Pull request merged**